### PR TITLE
[REF] point_of_sale,*pos*: remove Backbone.js, single loading request

### PR DESCRIPTION
**Pull request by pimodoo**
This commit accomplishes multiple objectives:

1. Loading of model data for the POS UI is now done in single
request (with some exemptions).
2. Backbone.js is removed and is replaced by reactivity.js
authored by the js framework team.
3. Better organization of assets in the `__manifest__.py`

Data Loading
------------

One request per model -- that's always been how the data is loaded
in POS UI. Now, we move the aggregation of data to the backend which
simplified the data loading. In the frontend, data loading is
initiated in `load_server_data` which makes the single rpc call
to the `load_pos_data` defined in `pos.session` model.

There remains some rpc calls during loading but the number is
greatly reduced.

This change result to faster loading of the POS UI. However,
there is no more progress bar during the loading.

Notable entry points:

- .js: `PosGlobalState.load_server_data`
- .py: `pos.session.load_pos_data`

It is possible to make customizations in the loading, but is done in
several steps:

1. Override `_pos_ui_models_to_load` to include the new model to load.
2. Define `_loader_params_<model_name>` to define the `search_params` and
optional `context`.
3. Define `_get_pos_ui_<model_name>` to return the data that will be
included in the full `loaded_data`. Perform the organization in this method
whenever necessary.
4. Override `_pos_data_process` for further post processing. The final
form of `loaded_data` in this method will be sent to the frontend.

Removal of Backbone.js
----------------------

POS is no longer dependent from `Backbone.js`, but it's replaced by a
new dependency -- `reactivity.js`. It's invented by the js framework
team which will be available in owl v2.

The consequences of this change are the following:

1. The old `PosModel` is renamed to `PosGlobalState`.
2. The name of the other models are kept (except Paymentline which
is renamed to Payment).
3. The name `PosModel` is now used as the base model of the data models.
This is symmetric to our use of `PosComponent`.
4. The instance of `PosGlobalState` is made reactive in `Chrome` (the
root component). As a result, whatever mutation made in that instance,
`Chrome` will rerender. But rendering is batched so multiple mutations
in a single sychronous function call will only result to a single
render call.
5. `PosModel`s can be extended in the `Registries` the same way as we
extend the `PosComponent`s.
```js
// E.g.
Registries.Component.extend(Chrome, PosResChrome);
Registries.Model.extend(Order, PosResOrder);
```

Reorganization of assets declaration
------------------------------------

This is a simple change that replaced the explicit enumeration
of loaded .js and .xml files. Also, we now have regions in the
manifest separating:

1. Augmentations of dependencies
2. Assets for `point_of_sale.index`
3. Assets for the qunit test (soon will be improved)

Other notable changes
---------------------

- Automatic call to `Order.save_to_db` is batched.
- Instances of `devices.DeviceProxy` (`proxy`), `devices.JobQueue`
(`proxy_queue`) and `BarcodeReader` (`barcode_reader`) are moved
to `point_of_sale.env`. Together with `posbus` and `posMutex`
(previously `flush_mutex`).
- `ActivePrograms` is now passed a props.
- `cashier` is no longer saved to `localStorage`.
- `pos_cache` is overhauled.
- `cache` is taken out from `PosDB` to prevent infinite rendering
loop whenever `cache` is mutated. It's now called `CACHE` in
the `db.js` file.
- reactive version of `posmodel` is exposed globally when in debug mode.
This means that calling when calling a mutator, the UI will react.
- Added control buttons in the `ProductScreen` is sorted during `Chrome`
setup. As a result, we don't need to worry on the loading order of the
control buttons components.
- Model can be simply instantiated using `BaseModel.create(obj)`. This
takes into account the extensions.
- `pos_coupon`: rewards are now updated whenever set_order is called.
- Component unit tests were removed. It wasn't useful and it just blocks
the development. We'll be introducing better unit testing later.

Co-authored-by: Jacky (trj) <trj@odoo.com>

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/l10n_in_pos/static/src/js/receipt.js [hunk 1]
```diff
@@ -2,16 +2,19 @@ odoo.define('l10n_in_pos.receipt', function (require) {
 "use strict";
 
 var models = require('point_of_sale.models');
+const Registries = require('point_of_sale.Registries');
 
-models.load_fields('product.product', 'l10n_in_hsn_code');
+Registries.PosModelRegistry.extend(models.Orderline, (Orderline) => {
```
Comment (pimodoo) [inline addons/l10n_in_pos/static/src/js/receipt.js:7]: @caburj This syntax is not very clear. Couldn't we have something that looks more like the ComponentRegistry syntax?

(Maybe after having merged the @trj-odoo PR)

Comment (caburj) [reply addons/l10n_in_pos/static/src/js/receipt.js:7]: Will do after merging TRJ's.

Comment (caburj) [reply addons/l10n_in_pos/static/src/js/receipt.js:7]: How about this?

https://github.com/odoo-dev/odoo/pull/1084
https://github.com/odoo-dev/enterprise/pull/227

File: addons/l10n_fr_pos_cert/static/src/js/pos.js [hunk 2]
```diff
@@ -23,38 +23,45 @@ models.PosModel = models.PosModel.extend({
...
       return _.contains(french_countries, this.company.country.code);
-    },
+    }
     disallowLineQuantityChange() {
-        let result = _super_posmodel.disallowLineQuantityChange.bind(this)();
+        let result = super.disallowLineQuantityChange(...arguments);
         return this.is_french_country() || result;
     }
-});
+    cashierHasPriceControlRights() {
```
Comment (caburj) [inline addons/l10n_fr_pos_cert/static/src/js/pos.js:31]: This is the replacement of removing NumpadWidget.js. It's used in computing `get hasPriceControlRights` in the main NumpadWidget component.

File: addons/l10n_co_pos/static/src/js/pos.js [hunk 3]
```diff
@@ -2,31 +2,33 @@ odoo.define('l10n_co_pos.pos', function (require) {
 "use strict";
 
 var models = require('point_of_sale.models');
+const Registries = require('point_of_sale.Registries');
 
-models.PosModel = models.PosModel.extend({
-    is_colombian_country: function () {
+const L10nCoPosGlobalState = (PosGlobalState) => class L10nCoPosGlobalState extends PosGlobalState {
```
Comment (pimodoo) [inline addons/l10n_co_pos/static/src/js/pos.js:7]: can you explain me the syntax of this I don't really understand it. Why do you need "const L10nCoPosGlobalState = (PosGlobalState) =>"?

File: addons/l10n_co_pos/static/src/js/pos.js [hunk 4]
```diff
@@ -2,31 +2,33 @@ odoo.define('l10n_co_pos.pos', function (require) {
...
+    is_colombian_country() {
         return this.company.country.code === 'CO';
-    },
-});
+    }
+}
+Registries.PosModelRegistry.extend(models.PosGlobalState, L10nCoPosGlobalState);
```
Comment (pimodoo) [inline addons/l10n_co_pos/static/src/js/pos.js:12]: Why do we need to explicitly call models.PosGlobalState, and not just "PosGlobalState", which is the class extended.

Comment (pimodoo) [reply addons/l10n_co_pos/static/src/js/pos.js:12]: Can't we just require the PosGlobalState

File: addons/l10n_co_pos/static/src/js/pos.js (hunk 3 above)
Comment (pimodoo) [reply addons/l10n_co_pos/static/src/js/pos.js:7]: forget this one, I get it

File: addons/l10n_fr_pos_cert/static/src/js/pos.js (hunk 2 above)
Comment (pimodoo) [reply addons/l10n_fr_pos_cert/static/src/js/pos.js:31]: ok

File: addons/l10n_fr_pos_cert/static/src/js/NumpadWidget.js [hunk 5]
```diff
@@ -1,20 +0,0 @@
-odoo.define('l10n_fr_pos_cert.NumpadWidget', function(require) {
-    'use strict';
-
```
[the remark this answers is not in the document]
Comment (caburj) [reply addons/l10n_fr_pos_cert/static/src/js/NumpadWidget.js:3]: https://github.com/odoo/odoo/pull/82461#discussion_r782850748

File: addons/l10n_co_pos/static/src/js/pos.js (hunk 4 above)
Comment (caburj) [reply addons/l10n_co_pos/static/src/js/pos.js:12]: We always need the base class when extending.

Comment (caburj) [reply addons/l10n_co_pos/static/src/js/pos.js:12]: > Can't we just require the PosGlobalState
yes, we can.

File: addons/point_of_sale/static/src/js/Screens/PaymentScreen/PaymentLine.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
+odoo.define('point_of_sale.PaymentLine', function(require) {
```
Comment (pimodoo) [inline addons/point_of_sale/static/src/js/Screens/PaymentScreen/PaymentLine.js:1]: Why doing this? it create a lot of useless diff, it just renames line in _line

Comment (caburj) [reply addons/point_of_sale/static/src/js/Screens/PaymentScreen/PaymentLine.js:1]: One of my early changes. I'll return the original component replaced by this.

File: addons/l10n_co_pos/static/src/js/pos.js (hunk 4 above)
Comment (caburj) [reply addons/l10n_co_pos/static/src/js/pos.js:12]: We can require `PosGlobalState` but it's not the extended version. It's still the base class. To get the extended version, we have to get it from the registry like so: `PosModelRegistry.get(PosGlobalState)`.

Do you want something such that when we do `const { PosGlobalState } = require('point_of_sale.models');`, `PosGlobalState` is already the extended version? I can't do that because when we need to have a way to get the base class so that we can extend it. However, I can do something like this:

```
const { PosGlobalState } = require('point_of_sale.models');

PosGlobalState.getBase() [or .Base] -> returns the base model which will be used for extending

PosGlobalState.getExtended() [or .Extended] -> returns the extended class which can be used for creating instances.
```

What do you think?

Comment (caburj) [reply addons/l10n_co_pos/static/src/js/pos.js]: Or do you mean I should do this? https://github.com/odoo/odoo/pull/82461/commits/fa2334540f12a687398dbf3e26946e3fe579e737

File: addons/point_of_sale/static/src/js/ChromeWidgets/ClientScreenButton.js [hunk 7]
```diff
@@ -29,16 +30,19 @@ odoo.define('point_of_sale.ClientScreenButton', function(require) {
             }
         }
         async onClickLocal() {
-            this.env.pos.customer_display = window.open('', 'Customer Display', 'height=600,width=900');
+            // Before, customer_display was in 'pos', which makes it available "globally".
+            // This time however, we are putting it in a different global place since
+            // we can't attached dom element in 'pos'.
+            pos_env.customer_display = window.open('', 'Customer Display', 'height=600,width=900');
```
Comment (pimodoo) [inline addons/point_of_sale/static/src/js/ChromeWidgets/ClientScreenButton.js:36]: Why do you have this pos_env stuff? In other components you have "this.env" that replaces "this.env.pos".

BTW, isn't it possible to keep "this.env.pos" to avoid useless diff in PR which is already very long?

Comment (pimodoo) [reply addons/point_of_sale/static/src/js/ChromeWidgets/ClientScreenButton.js:36]: BTW "this.env.pos" seems to be available as it is already used elsewhere

File: addons/point_of_sale/static/src/js/ChromeWidgets/ProxyStatus.js [hunk 8]
```diff
@@ -18,21 +18,10 @@ odoo.define('point_of_sale.ProxyStatus', function(require) {
             this.index = 0;
         }
         mounted() {
-            this.env.pos.proxy.on('change:status', this, this._onChangeStatus);
+            this.env.proxy.on('change:status', this, this._onChangeStatus);
         }
         willUnmount() {
-            this.env.pos.proxy.off('change:status', this, this._onChangeStatus);
-        }
-        async onClick() {
```
Comment (pimodoo) [inline addons/point_of_sale/static/src/js/ChromeWidgets/ProxyStatus.js:26]: Why did you remove this behavior?

File: addons/point_of_sale/static/src/js/ControlButtonsMixin.js [hunk 9]
```diff
@@ -11,22 +11,80 @@ odoo.define('point_of_sale.ControlButtonsMixin', function (require) {
      * @param {Function} x superclass
      */
     const ControlButtonsMixin = (x) => {
+        let sorted = false;
```
Comment (pimodoo) [inline addons/point_of_sale/static/src/js/ControlButtonsMixin.js:14]: Why moving logic here (out of "addControlButton"), it seems more complex and not needed

File: addons/point_of_sale/static/src/js/Popups/EditListInput.js [hunk 10]
```diff
@@ -13,7 +13,7 @@ odoo.define('point_of_sale.EditListInput', function(require) {
     }
     EditListInput.template = 'EditListInput';
 
-    Registries.Component.add(EditListInput);
+    Registries.PosComponentRegistry.add(EditListInput);
```
Comment (pimodoo) [inline addons/point_of_sale/static/src/js/Popups/EditListInput.js:16]: Finally, don't rename this registry, it makes a lot of noize.

Comment (pimodoo) [reply addons/point_of_sale/static/src/js/Popups/EditListInput.js:16]: You can also rename PosModelRegistry => Model

File: addons/point_of_sale/static/src/js/Screens/ProductScreen/ProductScreen.js [hunk 11]
```diff
@@ -161,13 +154,14 @@ odoo.define('point_of_sale.ProductScreen', function(require) {
             const { mode } = event.detail;
             NumberBuffer.capture();
             NumberBuffer.reset();
-            this.state.numpadMode = mode;
+            this.env.pos.PRODUCT_SCREEN.numpadMode = mode;
```
Comment (pimodoo) [inline addons/point_of_sale/static/src/js/Screens/ProductScreen/ProductScreen.js:157]: It doesn't seems to have other numpadMode for any other screen, so maybe you can store it in "this.env.pos.numpadMode"

File: addons/point_of_sale/static/src/js/ChromeWidgets/ClientScreenButton.js (hunk 7 above)
Comment (caburj) [reply addons/point_of_sale/static/src/js/ChromeWidgets/ClientScreenButton.js:36]: I'll remove this pos_env import in this component. But we need pos_env in the models as models are reactive and env will be become reactive when access thru pos. Making the env reactive is not a good idea (actually, pos won't work if we do so).

File: addons/point_of_sale/static/src/js/ControlButtonsMixin.js (hunk 9 above)
Comment (caburj) [reply addons/point_of_sale/static/src/js/ControlButtonsMixin.js:14]: You want to remove the declaration of the control buttons from the manifest so I removed it. But to maintain the ordering, I have to do the ordering when every popup component is added, thus, ordering shouldn't be made in addControlButton.

File: addons/point_of_sale/static/src/js/Popups/EditListInput.js (hunk 10 above)
Comment (caburj) [reply addons/point_of_sale/static/src/js/Popups/EditListInput.js:16]: Okay

File: addons/point_of_sale/static/src/js/Screens/ProductScreen/ProductScreen.js (hunk 11 above)
Comment (caburj) [reply addons/point_of_sale/static/src/js/Screens/ProductScreen/ProductScreen.js:157]: I can do that.

File: addons/point_of_sale/static/src/js/ControlButtonsMixin.js (hunk 9 above)
Comment (pimodoo) [reply addons/point_of_sale/static/src/js/ControlButtonsMixin.js:14]: Okay, I get it, I'll try to find if there is a prettier way to order it

File: addons/point_of_sale/static/src/js/ChromeWidgets/ProxyStatus.js (hunk 8 above)
Comment (caburj) [reply addons/point_of_sale/static/src/js/ChromeWidgets/ProxyStatus.js:26]: I didn't. Check the change in https://github.com/odoo/odoo/pull/82461/files/fbd966a132d6dd9c12fc7892192aa5714adfd61e#diff-e0849d9d6458581f1aeb9f69fcfa358761e1b1a775c37616dcb4dca20bf4a0e1

Comment (caburj) [reply addons/point_of_sale/static/src/js/ChromeWidgets/ProxyStatus.js:26]: Or check the change in `ProxyStatus.xml`

File: addons/point_of_sale/static/src/js/ControlButtonsMixin.js (hunk 9 above)
Comment (caburj) [reply addons/point_of_sale/static/src/js/ControlButtonsMixin.js:14]: Maybe we can sort the control buttons at setup of the Chrome component?

Comment (pimodoo) [reply addons/point_of_sale/static/src/js/ControlButtonsMixin.js:14]: Yes, you can move it there

**Comment by caburj** [comment]
@pimodoo I noticed that the biggest clutter now is the conversion of `this.env` to `env`. Do you want me to remove that commit to help simplify with your review? Then later we append the commit?

**Comment by pimodoo** [comment]
@caburj No, it is ok, you can let the env commit

**Comment by caburj** [comment]
@odoo/rd-security Hello. This PR needs some approval from your team. Regarding use of `html($`, that's already there, we are not actually introducing it so I guess you can easily approve it. What really needs attention is our use of `getattr`.

**Comment by mart-e** [comment]
> Regarding use of `html($`, that's already there, we are not actually introducing it so I guess you can easily approve it.

Not really, we have a lot of historical useless (and potentially dangerous) `.html()` calls so touching the line is a good occasion to evaluate if it can be removed or not. If not, is the content safe?

**Comment by caburj** [comment]
I'm not very knowledgable on this topic of security. The html string is generated in
```js
QWeb.render('CustomerFacingDisplayOrder', {
                pos: self,
                origin: window.location.origin,
                order: order,
                productImages
            })
```
Can you point me where to look to assess it's safety?

File: addons/point_of_sale/static/src/js/models.js [hunk 12]
```diff
@@ -522,7 +522,7 @@ class PosGlobalState extends PosModel {
     }
 
     send_current_order_to_customer_facing_display() {
-        if (!this.config.iface_customer_facing_display) return;
```
Comment (pimodoo) [inline addons/point_of_sale/static/src/js/models.js]: Are you sure we don't want to send to customer display when no order? Because when you are finalizing an order and then you don't have one anymore, we should "clear" the screen. (or maybe we always have an empty order, so in which case don't we have an order?

Comment (caburj) [reply addons/point_of_sale/static/src/js/models.js]: True true.

**Comment by pimodoo** [comment]
@trj-odoo @caburj Sorry for the conflict on giftcard, I forwardported a fix. We don't need to load anymore the giftcard when we load the POS.

**Comment by caburj** [comment]
@pimodoo Okay. I'll resolve it.

**Comment by caburj** [comment]
@mart-e Can you help me in assessing whether the html string is safe or not?

File: addons/point_of_sale/models/pos_session.py [hunk 13]
```diff
@@ -1550,6 +1553,425 @@ def try_cash_in_out(self, _type, amount, reason, extras):
             message_content.append(f'- Reason: {reason}')
         self.message_post(body='<br/>\n'.join(message_content))
 
+    def get_onboarding_data(self):
+        return {
+            "categories": self._load_model('pos.category'),
+            "products": self._load_model('product.product'),
+        }
+
+    def get_loading_params(self, model):
```
[the remark this answers is not in the document]
Comment (caburj) [reply addons/point_of_sale/models/pos_session.py:1562]: You're right. I'll do that.

[the remark this answers is not in the document]
Comment (trj-odoo) [reply addons/point_of_sale/models/pos_session.py:1562]: @caburj, Robin made a `get_pos_ui_res_partner_by_params` method for his fix, maybe you could use that as well

[the remark this answers is not in the document]
Comment (caburj) [reply addons/point_of_sale/models/pos_session.py:1562]: ah yes, thanks

[the remark this answers is not in the document]
Comment (caburj) [reply addons/point_of_sale/models/pos_session.py]: can you check the latest commit?

[the remark this answers is not in the document]
Comment (trj-odoo) [reply addons/point_of_sale/models/pos_session.py]: That should work

**Comment by ivantodorovich** [comment]
🙏🏻 thanks!!!
This is great, backbone was so outdated

**Comment by xmo-odoo** [comment]
@robodoo override=ci/security

`getattr` on *partially* customisable attrnames

File: addons/point_of_sale/models/pos_session.py [hunk 14]
```diff
@@ -1550,6 +1553,419 @@ def try_cash_in_out(self, _type, amount, reason, extras):
...
+    def get_onboarding_data(self):
+        return {
+            "categories": self._load_model('pos.category'),
+            "products": self._load_model('product.product'),
+        }
+
+    def _load_model(self, model):
+        model_name = model.replace('.', '_')
+        if hasattr(self, '_get_pos_ui_%s' % model_name) and hasattr(self, '_loader_params_%s' % model_name):
+            return getattr(self, '_get_pos_ui_%s' % model_name)(getattr(self, '_loader_params_%s' % model_name)())
```
Comment (xmo-odoo) [inline addons/point_of_sale/models/pos_session.py:1565]: Seems unnecessary to duplicate the lookup (and formatting), could be e.g.
```suggestion
        loader = getattr(self, '_get_pos_ui_%s' % model_name, None)
        params = getattr(self, '_loader_params_%s' % model_name, None)
        if loader and params:
            return loader(params())
```

**Comment by mart-e** [comment]
> Can you help me in assessing whether the html string is safe or not?

@caburj sorry for the late reply. I see you have managed to avoid touching these lines. Indeed when it is the result of a qweb template, it is _usually_ fine (as long as you properly escape the content of the template).

**Comment by pimodoo** [comment]
@robodoo r+ rebase-ff

**Comment by pimodoo** [comment]
@robodoo r+ rebase-ff

**Comment by pimodoo** [comment]
@robodoo retry

File: addons/account/models/account_tax.py [hunk 15]
```diff
@@ -105,6 +105,7 @@ def _default_tax_group(self):
     sequence = fields.Integer(required=True, default=1,
         help="The sequence field is used to define order in which the tax lines are applied.")
     amount = fields.Float(required=True, digits=(16, 4), default=0.0)
+    real_amount = fields.Float(string='Real amount to apply', compute='_compute_real_amount', store=True)
```
Comment (smetl) [inline addons/account/models/account_tax.py:108]: This field is useless....
- Used only for POS but can be replaced by a single multiplication
- Part of no business flow
- No review asked to the accounting team
- No migration script for that
Come on guys...

Comment (pimodoo) [reply addons/account/models/account_tax.py:108]: It was here as a fix in a function since 2019 https://github.com/odoo/odoo/pull/36140, because you changed the taxes and broke the POS tax computation with your repartition lines. It is used in front and was set as a field to get it easily when we need to compute tax.

Just missing the migration script, my bad, but all other reasons are wrong, and it does not impact the accounting stuff so I don't really see why we would have need your opinion.

I doubt that all your migration scripts are there and perfect too.

Comment (trj-odoo) [reply addons/account/models/account_tax.py:108]: > Part of no business flow

Just FYI, it is sent to and used in the PoS UI during the loading thanks to `load_pos_data` and `_pos_ui_models_to_load`. Hence why you couldn't find any occurrences of  `_get_pos_ui_account_tax`

Comment (pimodoo) [reply addons/account/models/account_tax.py:108]: @smetl BTW we should probably use it on the sale orders and the invoices to avoid having those kind of inconsistencies when the repartition lines are not fullfulling the amount.
 
 
image

Comment (smetl) [reply addons/account/models/account_tax.py:108]: @pimodoo The repartition lines are well managed python-side. This is absolutely not related to the unconsistency with sale orders.
This field is useless because you can perform yourself the computation of your `real_amount` (that doesn't mean anything btw...) in `_get_pos_ui_account_tax`.
And sorry but you **need** to ask a review for that. We are challenged for any single field we add. This is not up to you to decide that yourself.

Comment (pimodoo) [reply addons/account/models/account_tax.py:108]: The real amount, is the effective percentage that you have to apply on your base price to know your total price.
